### PR TITLE
feat(agent_api_keys): one-call webhook-trigger setup endpoint

### PR DESCRIPTION
**Pull request by danielmillerp**
## What
Adds `POST /agent_api_keys/webhook-trigger` — wires a webhook trigger in **one call**:
- registers a `github`/`slack` signature-verification key for the agent (auto-generates the signing secret if not provided),
- returns the ready-to-paste **forward webhook URL** + the **secret** (shown once).

```
POST /agent_api_keys/webhook-trigger
{ "agent_name": "golden-agent", "source": "github",
  "name": "<owner/repo>", "forward_path": "github-pr/<config-id>" }
→ { key_id, secret, webhook_path, webhook_url }
```

## Why
The pieces to trigger an agent from a webhook already exist (the `/agents/forward` ingress verifies the signature against an agent key, and `POST /agent_api_keys` registers keys). This bundles key-create + webhook-URL composition so a UI (or a curl) can set up a trigger in a single step instead of two — the backend for the self-serve "Add trigger" button. The webhook then flows through the existing forward ingress unchanged.

**Before** — wiring a trigger was two manual steps, and the caller had to know the agent's internal id, invent a secret, and hand-compose the forward URL:

```python
# 1. register the signing key (need the agent's internal id + your own secret)
POST /agent_api_keys
{ "agent_id": "<look-up-first>", "api_key": "<generate-yourself>",
  "name": "owner/repo", "api_key_type": "github" }

# 2. hand-build the forward URL from the convention you have to know
webhook_url = f"{public_url}/agents/forward/name/{agent_name}/github-pr/{config_id}"
```

**After** — one call, by agent **name**, returns the URL + secret ready to paste:

```python
POST /agent_api_keys/webhook-trigger
{ "agent_name": "golden-agent", "source": "github",
  "name": "owner/repo", "forward_path": "github-pr/<config-id>" }

→ { "key_id": "...", "secret": "ab3f…",            # generated for you, shown once
    "webhook_path": "/agents/forward/name/golden-agent/github-pr/<config-id>",
    "webhook_url":  "https://<host>/agents/forward/name/golden-agent/github-pr/<config-id>" }
```

No new ingress, no migration — built on the existing `agent_api_keys` + forward mechanism.

## On the signing secret
This matches how GitHub webhooks actually work. A GitHub webhook's **Secret** is a value *you* supply (GitHub doesn't generate one) — you type a high-entropy string into the webhook's Secret field, and GitHub uses it to HMAC each payload into `X-Hub-Signature-256: sha256=…`, which the receiver re-computes and compares. It's a shared secret that must exist on both sides (GitHub's config + the verifying server), and it's optional-but-recommended (no secret → no signature header at all). Refs: GitHub docs [validating-webhook-deliveries](https://docs.github.com/en/webhooks/using-webhooks/validating-webhook-deliveries), [creating-webhooks](https://docs.github.com/en/webhooks/using-webhooks/creating-webhooks).

Given that, the endpoint's `secret` handling is deliberate:
- **omit `secret`** (default) → the endpoint generates a strong one (`secrets.token_hex(32)`), stores it on the agent's verification key, and returns it once to paste into GitHub's Secret field — so the caller never has to invent entropy.
- **supply `secret`** → for when the GitHub webhook already has a secret set and the agent side just needs to match it.

## Testing
- 4 unit tests (URL composition, provided vs generated secret, 409 on duplicate, 400 on non-webhook source). ruff clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR adds `POST /agent_api_keys/webhook-trigger`, a single-call convenience endpoint that registers a GitHub or Slack signature-verification key for an agent and returns the ready-to-paste forward webhook URL and signing secret. It builds on the existing `agent_api_keys` store and `/agents/forward` ingress without any new migration or routing.

- The handler correctly validates source type, enforces Slack callers supply their own Signing Secret (vs. auto-generating for GitHub), URL-encodes agent name and forward path via `urllib.parse.quote`, and checks for control characters in `forward_path` before the DB write — all issues from earlier review rounds have been addressed.
- Eight unit tests cover URL composition, provided vs. generated secret, empty-string secret preservation, URL-encoding of reserved characters, control-char rejection, duplicate-key 409, invalid-source 400, and Slack-without-secret 400.

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — all validation (source check, Slack secret requirement, control-char rejection) runs before the DB write, URL components are correctly percent-encoded, and the 409 response no longer leaks internal identifiers.

All issues raised in earlier review rounds have been addressed: input validation precedes the key creation call, the is not None guard preserves empty-string secrets without silent coercion, Slack requires the caller to supply the Signing Secret, and the forward path is properly encoded. The unit tests exercise all meaningful branches including the fixed edge cases.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| agentex/src/api/routes/agent_api_keys.py | New create_webhook_trigger handler with correct validation ordering (all input checks precede the DB write), explicit None-check for secret, URL-encoding for agent name and forward path, and a clean 409 message that avoids leaking internal UUIDs. |
| agentex/src/api/schemas/agent_api_keys.py | Adds CreateWebhookTriggerRequest and CreateWebhookTriggerResponse Pydantic models; secret field description now accurately documents the Slack requirement (must be supplied) and GitHub behaviour (auto-generated if omitted). |
| agentex/tests/unit/api/test_webhook_trigger.py | Eight focused unit tests covering happy path, provided/generated/empty secrets, URL-encoding of reserved characters, control-char rejection (with create mock asserting it was never called), duplicate 409, invalid source 400, Slack-without-secret 400, and Slack-with-secret success. |
| agentex/openapi.yaml | Adds the new endpoint and both request/response schemas; documents 200 and 422 consistent with all existing endpoints in the file. |

</details>

</details>

<sub>Reviews (7): Last reviewed commit: ["Merge branch &#39;main&#39; into dm/agentex-webh..."](https://github.com/scaleapi/scale-agentex/commit/b3b4ecde8c87169de7d8f3c87f7112756ea30ff2) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38966389)</sub>

### CODE REVIEW

File: agentex/src/api/routes/agent_api_keys.py [hunk 1]
```diff
@@ -93,6 +96,74 @@ async def create_api_key(
...
+
+    existing_api_key = await agent_api_key_use_case.get_by_agent_id_and_name(
+        agent_id=agent.id,
+        name=request.name,
+        api_key_type=request.source,
+    )
+    if existing_api_key:
+        error_msg = f"{request.source} webhook key '{request.name}' already exists for agent ID {agent.id}."
+        logger.error(error_msg)
+        raise HTTPException(status_code=409, detail=error_msg)
```
[the remark this answers is not in the document]
Comment (danielmillerp) [reply agentex/src/api/routes/agent_api_keys.py]: Fixed — dropped the `logger.error` for the 409 (a duplicate is an expected client condition, not a server error).

[the remark this answers is not in the document]
Comment (danielmillerp) [reply agentex/src/api/routes/agent_api_keys.py]: Fixed — the 409 detail no longer includes the internal agent UUID; it references only the key name the caller supplied.

File: agentex/src/api/routes/agent_api_keys.py [hunk 2]
```diff
@@ -93,6 +96,77 @@ async def create_api_key(
...
+) -> CreateWebhookTriggerResponse:
+    """Wire a webhook trigger in one call.
+
+    Registers the source's signature-verification key (github/slack) for the agent and
+    returns the ready-to-paste forward webhook URL plus the signing secret (shown once).
+    The webhook then flows through the existing /agents/forward ingress, which verifies
+    the signature against this key. Bundles the existing key-create + URL composition so
+    a UI (or a curl) can set up a trigger without two steps.
+    """
+    if request.source not in (AgentAPIKeyType.GITHUB, AgentAPIKeyType.SLACK):
```
Comment (declan-scale) [inline agentex/src/api/routes/agent_api_keys.py:122]: Are there other webhooks we want to scope out (linear, notion, pager duty, datadog?) Is there a way to make it easier to add a key than updating the code?

**Review by declan-scale — APPROVED** [review]

**Comment by declan-scale** [comment]
Will we want a rotate path as well should we want to rotate the api key?

File: agentex/src/api/routes/agent_api_keys.py (hunk 2 above)
Comment (danielmillerp) [reply agentex/src/api/routes/agent_api_keys.py:122]: Deferring to follow-ups, keeping this PR scoped to github/slack. Each new source (Linear, Notion, PagerDuty, Datadog) needs its own signature-verification scheme wired into the forward path — Linear is HMAC-SHA256 like GitHub, but PagerDuty/Datadog/Notion differ — so it's real per-source work rather than just an enum value. When we add the next source we'll introduce a small `{source: signature_scheme}` registry so adding one becomes a single entry instead of scattered changes. Tracking that as the extension point.

File: agentex/src/api/routes/agent_api_keys.py [hunk 3]
```diff
@@ -93,6 +96,77 @@ async def create_api_key(
...
+            status_code=409,
+            detail=f"A {request.source} webhook key named '{request.name}' already exists for this agent.",
+        )
+
+    secret = request.secret or secrets.token_hex(32)
+    agent_api_key_entity = await agent_api_key_use_case.create(
+        agent_id=agent.id,
+        api_key=str(secret),
+        name=request.name,
+        api_key_type=request.source,
```
[the remark this answers is not in the document]
Comment (danielmillerp) [reply agentex/src/api/routes/agent_api_keys.py]: Fixed in the latest commit — Slack triggers now require the caller to supply `secret` (the app's Signing Secret); a missing secret 400s instead of storing a generated value that could never match Slack's signature. GitHub still auto-generates (that one you do paste into the repo). Added unit tests for both the Slack-without-secret 400 and the Slack-with-secret happy path.

**Comment by danielmillerp** [comment]
> Will we want a rotate path as well should we want to rotate the api key?

Good call — deferring a dedicated rotate path to a follow-up, keeping this PR to the create flow. Today rotation works by **delete + recreate**: the forward URL/path is derived from agent name + forward_path (not the secret), so deleting the key and re-creating the trigger yields a new secret to paste while the webhook URL stays the same. A dedicated `rotate` (regenerate-secret-in-place, returned once, URL unchanged) is cleaner and avoids the brief gap between delete and recreate — worth adding alongside the M2M owner-key work where key rotation matters more. Tracking it as a follow-up.

**Comment by declan-scale** [comment]
@danielmillerp can you address the greptile comments, then good to go

**Comment by danielmillerp** [comment]
Addressed the latest Greptile comments in de85c65.

- Preserves an explicitly provided empty GitHub secret by using an explicit None check instead of truthiness.
- URL-encodes agent_name and forward_path when composing the returned forward URL, while preserving subpath slashes.
- Rejects control characters in forward_path.
- Updates the request schema docs so Slack clearly requires the app Signing Secret, while GitHub may omit secret to generate one.
- Added unit coverage for empty GitHub secret, reserved URL chars, control-char rejection, and the Slack secret schema text.
